### PR TITLE
ball transport subsystem

### PR DESCRIPTION
**Pull request by nmot919**
Adds ball transport subsystem
Adds automatic ball shunting
Adds balls tracking

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            bottomMotor.set(0);
+        });
+    }
+
+    public CommandBase shuntWrongCargo() {
+        return cmd("Shunt Wrong Cargo").onExecute(() -> {
+            switch (allianceColor) {
+                case Red:
+                    if (colorBuffer.peekLast() == Color.kFirstBlue) {
+                        topMotor.set(SHUNT_SPEED);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: This needs to run the shooter as well

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                        colorBuffer.removeLast();
+                    }
+                    if (colorBuffer.peekFirst() == Color.kFirstBlue) {
+                        bottomMotor.set(-SHUNT_SPEED);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: If we had to get rid of the first ball through the shooter we should probably do the same with the second ball.
If we really need to reverse it out the intake than we need to deploy the intake first.

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        }).finishedWhen(() -> {
+            return cargoInColorSensor && laserSwitch.getAsBoolean();
+        }).onEnd(() -> {
+            topMotor.set(0);
+            bottomMotor.set(0);
+        });
+    }
+
+    public CommandBase shuntWrongCargo() {
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: We probably need to pull parts of this logic out into robot.java and create sequential groups with the shooter and intake. (Or access those subsystems from here, I'm not sure which is uglier)

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                        bottomMotor.set(-SHUNT_SPEED);
+                        colorBuffer.removeFirst();
+                    }
+                    break;
+                case Invalid:
+                default:
+                    break;
+            }
+        }).finishedWhen(() -> {
+            return !cargoInColorSensor || !laserSwitch.getAsBoolean();
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: What if both cargo were the wrong color? This might end the command too soon.

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        }
+
+    }
+
+    @Override
+    public void periodic() {
+        updateColorBuffer();
+        SmartDashboard.putBoolean("Laser Switch Activated", laserSwitch.getAsBoolean());
+        SmartDashboard.putBoolean("Cargo in Color Sensor", cargoInColorSensor);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: It would be good to send the color of any/all balls we have currently. See [this](https://docs.wpilib.org/en/stable/docs/software/dashboards/shuffleboard/layouts-with-code/organizing-widgets.html) for an idea on how to make this look good.

**Review by bot190 — CHANGES_REQUESTED** [review]
Can you add an implementation of `getBallTransportMap` to `GastonMap.java`?

We'll definitely have to spend some time tweaking `shuntWrongCargo` as there's lots of room for optimization here.

Probably more importantly is this needs a command to deliver a ball to the shooter!

File: src/main/java/frc/robot/subsystems/BallTransport.java (hunk 3 above)
Comment (nmot919) [reply src/main/java/frc/robot/subsystems/BallTransport.java]: personally id want to import the shoot and intake to the balltransport just to make life easier but as of now shooter code hasnt been pushed to main so we'll need to wait on that one.

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    private final IColorSensor colorSensor;
+    private final BooleanSupplier laserSwitch;
+    private final Alliance allianceColor;
+
+    private boolean cargoInColorSensor = false;
+    private final double SHUNT_SPEED = 1.0;
+    private final double TRANSPORT_SPEED = 1.0;
+
+    // creates a color sample buffer.
+    private SampleBuffer<Color> colorBuffer = new SampleBuffer<Color>(2);
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: Pro tip - you can actually remove the generic specialization on the right hand side

```java
private SampleBuffer<Color> colorBuffer = new SampleBuffer<>(2);
```

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            return cargoInColorSensor && laserSwitch.getAsBoolean();
+        }).onEnd(() -> {
+            topMotor.set(0);
+            bottomMotor.set(0);
+        });
+    }
+
+    public CommandBase shuntWrongCargo() {
+        return cmd("Shunt Wrong Cargo").onExecute(() -> {
+            switch (allianceColor) {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: You have duplicate logic in each case of the switch, the only difference is the colors. Why not start the command by determining which color is Our Color and which is Other Color? Then you can use the same motor logic, just comparing based on color variables.

Another issue is that `allianceColor` is set once when the subsystem is created. This is when the robot is turned on. The problem is that we don't actually get the alliance color at that point, we only get it once we're connected to the field. So this variable shouldn't be stored, it should be accessed when it's needed.

File: src/main/java/frc/robot/subsystems/BallTransport.java (hunk 3 above)
Comment (msoucy) [reply src/main/java/frc/robot/subsystems/BallTransport.java]: "Import the shoot and intake to the ball transport"? 🚩 

The main purpose of subsystems is to separate the logical concerns. Anything that relies on more than one subsystem should be done in Robot.java as a command group.

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                } else {
+                    bottomMotor.set(TRANSPORT_SPEED);
+                }
+            } else {
+                bottomMotor.set(TRANSPORT_SPEED);
+                topMotor.set(TRANSPORT_SPEED);
+            }
+        }).finishedWhen(() -> {
+            return cargoInColorSensor && laserSwitch.getAsBoolean();
+        }).onEnd(() -> {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: You can use `this::stop` instead of the lambda that just repeats `safeState`

**Review by msoucy — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/BallTransport.java (hunk 6 above)
Comment (jeffaburke) [reply src/main/java/frc/robot/subsystems/BallTransport.java]: This was fixed

File: src/main/java/frc/robot/subsystems/BallTransport.java (hunk 8 above)
Comment (jeffaburke) [reply src/main/java/frc/robot/subsystems/BallTransport.java]: This was implemented

File: src/main/java/frc/robot/Robot.java [hunk 9]
```diff
@@ -37,8 +41,10 @@ public void configureButtonBindings() {
 
     driveController.back().whenPressed(drive.resetCmd());
 
-    copilotController.a().whileHeld(intake.extend(SpinDirection.COUNTERCLOCKWISE));
-    copilotController.b().whileHeld(intake.retract(SpinDirection.COUNTERCLOCKWISE));
+    copilotController.a().whileHeld(parallel("Intake extended w/ Ball Transport",
+        intake.extend(SpinDirection.COUNTERCLOCKWISE), ballTransport.loadCargoWithIntake()));
+    copilotController.b().whileHeld(parallel("Intake retracted w/ Ball Transport",
+        intake.retract(SpinDirection.COUNTERCLOCKWISE), ballTransport.loadCargoNoIntake()));
```
Comment (bot190) [inline src/main/java/frc/robot/Robot.java]: This will have to be merged with the latest intake code which only has one mapping here.

File: src/main/java/frc/robot/maps/RobotMap.java [hunk 10]
```diff
@@ -136,6 +187,10 @@ public BooleanSupplier getOutsideLimit() {
         }
     }
 
+    public IntakeMap getIntakeMap() {
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/RobotMap.java:176]: Why was this moved?

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+import frc.robot.Robot;
+import frc.robot.maps.RobotMap.BallTransportMap;
+
+public class BallTransport extends SmartSubsystemBase {
+    private final SmartMotorController bottomMotor;
+    private final SmartMotorController topMotor;
+    private final IColorSensor colorSensor;
+    private final BooleanSupplier laserSwitch;
+
+    static final private int BALL_DETECTION_LIMIT = 1000;
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: This should have a comment that it's a distance in some arbitrary units.

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        return colorSensor.getProximity() < BALL_DETECTION_LIMIT;
+    }
+
+    public boolean getLaserSwitch() {
+        return laserSwitch.getAsBoolean();
+    }
+
+    // command selector enum used for command selector
+    public enum CommandSelector {
+        COLORZEROLASERZERO,
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: Can these names be made more descriptive at all?

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+                bottomMotor.stopMotor();
+            }
+        }).until(() -> {
+            return laserSwitch.getAsBoolean();
+        });
+    }
+
+    // select command that determines what command needs to be run based on the
+    // command selector.
+    public SelectCommand loadCargoWithIntake() {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java:110]: Please keep this with the appropriate variables

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+        return new SelectCommand(selectCommandMap, this::commandSelector);
+    }
+
+    // TODO Needs to be fixed
+    // Unloads cargo that is opposite of the alliance color
+    // Intake must be deployed backwards for this
+    /*
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: Commented out code?

**Review by msoucy — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+
+    public BallTransport(final BallTransportMap map) {
+        this.bottomMotor = map.getBottomMotor();
+        this.topMotor = map.getTopMotor();
+        this.colorSensor = map.getColorSensor();
+        this.laserSwitch = map.getLaserSwitch();
+    }
+
+    private boolean colorSensorBallLimit() {
+        return colorSensor.getProximity() < BALL_DETECTION_LIMIT;
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: If the proximity result gets bigger as things get closer, is `<` the right operation here?

File: src/main/java/frc/robot/subsystems/BallTransport.java [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+    public SelectCommand loadCargoWithIntake() {
+        return new SelectCommand(selectCommandMap, this::commandSelector);
+    }
+
+    // Unloads cargo that is opposite of the alliance color
+    // Intake must be deployed backwards for this
+    public CommandBase removeCargo() {
+        return cmd("Remove \"Wrong Colored\" Cargo").onExecute(() -> {
+            switch (allianceColor) {
+                default:
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/BallTransport.java]: `default` case should always come after the other cases - the `switch` goes into the first branch that it matches. When that is `default`, then everything will trigger that case.

File: src/main/java/frc/robot/maps/RobotMap.java (hunk 10 above)
Comment (jeffaburke) [reply src/main/java/frc/robot/maps/RobotMap.java:176]: It was at the bottom of the file so I moved it right below the intake map

**Review by msoucy — APPROVED** [review]

**Review by bot190 — APPROVED** [review]
